core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java

```diff
@@ -248,33 +250,76 @@ protected void doCheck(Object key, boolean isWrite, long flagBitSet) {
-      boolean operationAllowed = isOperationAllowed(isWrite, flagBitSet);
-      if (!actualMembers.containsAll(owners) && !operationAllowed) {
+      boolean operationAllowed = isKeyOperationAllowed(isWrite, flagBitSet, cacheTopology, key);
+      if (!operationAllowed) {
          if (trace) log.tracef("Partition is in %s mode, PartitionHandling is set to to %s, access is not allowed for key %s", availabilityMode, partitionHandling, key);
-         throw log.degradedModeKeyUnavailable(key);
+         if (EnumUtil.containsAny(flagBitSet, FlagBitSets.FORCE_WRITE_LOCK)) {
+            throw log.degradedModeLockUnavailable(key);
+         } else {
+            throw log.degradedModeKeyUnavailable(key);
+         }
       } else {
          if (trace) log.tracef("Key %s is available.", key);
       }
    }
 
-   protected boolean isOperationAllowed(boolean isWrite, long flagBitSet) {
+   /**
+    * Check if a read/write operation is allowed with the actual members
+    *
+    * @param isWrite       {@code false} for reads, {@code true} for writes
+    * @param flagBitSet    reads with the {@link org.infinispan.context.Flag#FORCE_WRITE_LOCK} are treated as writes
+    * @param cacheTopology actual members, or {@code null} for bulk operations
+    * @param key           key owners, or {@code null} for bulk operations
+    * @return
+    */
+   protected boolean isKeyOperationAllowed(boolean isWrite, long flagBitSet,
+                                           LocalizedCacheTopology cacheTopology, Object key) {
       if (availabilityMode == AvailabilityMode.AVAILABLE)
          return true;
 
+      List<Address> actualMembers = cacheTopology.getActualMembers();
       switch (partitionHandling) {
          case ALLOW_READ_WRITES:
             return true;
          case ALLOW_READS:
-            if (EnumUtil.containsAny(flagBitSet, FlagBitSets.FORCE_WRITE_LOCK))
-               throw log.degradedModeLockUnavailable();
-            return !isWrite;
+            List<Address> owners = getOwners(cacheTopology, key, isWrite);
+            if (isWrite || EnumUtil.containsAny(flagBitSet, FlagBitSets.FORCE_WRITE_LOCK)) {
+               // Writes require all the owners to be in the local partition
+               return actualMembers.containsAll(owners);
+            } else {
+               // Reads only require one owner in the local partition
+               return InfinispanCollections.containsAny(actualMembers, owners);
+            }
+         default:
+            // Both reads and writes require all the owners to be in the local partition
+            return actualMembers.containsAll(getOwners(cacheTopology, key, isWrite));
+      }
+   }
+
+   protected boolean isBulkOperationAllowed() {
+      if (availabilityMode == AvailabilityMode.AVAILABLE)
+         return true;
+
+      switch (partitionHandling) {
+         case ALLOW_READ_WRITES:
+            return true;
+         case ALLOW_READS:
+            // Never allow bulk reads/writes
```

Comment:
Why not allowing bulk reads? (please add comment)

Reply:
Good question... I think we can allow bulk reads for replicated caches :)